packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -90,6 +90,26 @@ export function useNumberFieldState(
     isReadOnly
   } = props;
 
+  if (value === null) {
+    value = NaN;
+  }
+
+  if (!isNaN(value)) {
+    if (!isNaN(step)) {
+      value = snapValueToStep(value, minValue, maxValue, step);
+    } else {
+      value = clamp(value, minValue, maxValue);
+    }
+  }
```

Comment:
so this is a departure from native behavior where the initial value provided to the input is respected even if it doesn't match the step but I saw in the spec where it recommends this new behavior so I'm fine with this, feels better anyways